device/simulation/tt_simulation_host.cpp

```diff
@@ -0,0 +1,72 @@
+// SPDX-FileCopyrightText: (c) 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <typeinfo>
+#include <sstream>
+#include <iomanip>
+#include <filesystem>
+#include <cassert>
+
+#include "common/logger.hpp"
+#include "common/assert.hpp"
+#include "tt_simulation_host.hpp"
+
+tt_SimulationHost::tt_SimulationHost() {
+    // Open socket and create dialer
+    nng_pair1_open(&host_socket);
+    std::string nng_socket_addr_str;
+    const char *ipc_socket_name = std::getenv("NNG_SOCKET_NAME");
+    if (!ipc_socket_name) {
+        nng_socket_addr_str = std::string(NNG_SOCKET_PREFIX) + std::string("nng_ipc_socket");
+        log_info(tt::LogEmulationDriver, "NNG_SOCKET_NAME not set, defaulting to 'ipc:///tmp/nng_ipc_socket'");
+    }else {
+        nng_socket_addr_str = std::string(NNG_SOCKET_PREFIX) + std::string(ipc_socket_name);
+    }
+    const char *nng_socket_addr = nng_socket_addr_str.c_str();
+
+    log_info(tt::LogEmulationDriver, "Dialing: {}", nng_socket_addr);
+    int rv = nng_dialer_create(&host_dialer, host_socket, nng_socket_addr);
+    TT_ASSERT(rv == 0, "Failed to create dialer: {} {}", nng_strerror(rv), nng_socket_addr);
+}
+
+tt_SimulationHost::~tt_SimulationHost() {
+    nng_dialer_close(host_dialer);
+    nng_close(host_socket);
+}
+
+void tt_SimulationHost::start_host() {
+    // Establish connection with remote VCS simulator
+    int rv;
+    do {
+        rv = nng_dialer_start(host_dialer, 0);
+        if (rv != 0) {
+            log_info(tt::LogEmulationDriver, "Waiting for remote: {}", nng_strerror(rv));
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+        }
+    } while (rv != 0);
+}
+
+void tt_SimulationHost::send_to_device(uint8_t *buf, size_t buf_size) {
+    int rv;
+    log_debug(tt::LogEmulationDriver, "Sending messsage to remote..");
+    
+    void *msg = nng_alloc(buf_size);
```

Comment:
Curious if this alloc and copy is necessary?  (Can `buf` get passed into `nng_send` directly?)

Reply:
I think doing so with the NNG provided alloc is safer since NNG will then manage the memory space and ensure compatibility with `nng_send/recv`. `nng_send` will also free the memory after sending the message. 

This should prevent any weird behaviour when sending data over the network